### PR TITLE
Make optional options optional in typings and add type tests

### PR DESCRIPTION
**Pull request by jdforsythe**
# Purpose / Goal

See #612 

This changes the type definitions to set all options with defaults to optional in the typings.

It adds documentation to all types so they show up in Intellisense when using the types.

It adds a type test - defining objects as the various options types (parse, validate, build) with none of the optional properties. Simply run `npm test-types` to attempt to compile the TS file and it will emit any type errors.

Example with no errors:

```sh
$ npm run test-types

> fast-xml-parser@4.3.2 test-types
> tsc --noEmit spec/typings/typings-test.ts

$ echo $?
0
## exit code 0 - no error
```

Example with an error - add `someNewProperty` to the type definition as non-optional:

fxp.d.ts:

```ts
type validationOptions = {
  /**
   * Whether to allow attributes without value
   * 
   * Defaults to `false`
   */
  allowBooleanAttributes?: boolean;
  
  /**
   * List of tags without closing tags
   * 
   * Defaults to `[]`
   */
  unpairedTags?: string[];

  // ADD A NEW PROPERTY - FORGET TO MAKE IT OPTIONAL
  /**
   * Some new property
   *
   * Defaults to `false`
   */
  someNewProperty: boolean;
};
```

The type test will now fail since the `typings-test.ts` file does not include this property in the validation options and it's listed as required:

```sh
$ npm run test-types

> fast-xml-parser@4.3.2 test-types
> tsc --noEmit spec/typings/typings-test.ts

spec/typings/typings-test.ts:41:7 - error TS2741: Property 'someNewProperty' is missing in type '{}' but required in type 'validationOptions'.

41 const validateOpts: validationOptions = {};
         ~~~~~~~~~~~~

  src/fxp.d.ts:234:2
    234   someNewProperty: boolean;
          ~~~~~~~~~~~~~~~~~~~~~~
    'someNewProperty' is declared here.


Found 1 error in spec/typings/typings-test.ts:41

$ echo $?
2
## exit code 2 - error
```

This will ensure the test fails if a property is added to the type as non-optional but not added to the `typings-test.ts` file as a required property.

# Type
Please mention the type of PR

* [ ]Bug Fix
* [ ]Refactoring / Technology upgrade
* [ ]New Feature

**Note** : Please ensure that you've read contribution [guidelines](https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/CONTRIBUTING.md) before raising this PR. If your PR is in progress, please prepend `[WIP]` in PR title. Your PR will be reviewed when `[WIP]` will be removed from the PR title.

[Bookmark](https://github.com/NaturalIntelligence/fast-xml-parser/stargazers) this repository for further updates.

### CODE REVIEW

File: package-lock.json
Comment (amitguptagwl) [inline package-lock.json:1]: Can you please remove this file from the PR?

**Comment by amitguptagwl** [comment]
Hi @jdforsythe 
Yarn.lock or package-lock both files can't be considered in a PR due to security reasons so they need to be removed from the PR.

**Comment by jdforsythe** [comment]
@amitguptagwl this has been fixed

**Review by amitguptagwl — APPROVED** [review]

**Comment by amitguptagwl** [comment]
Thanks for this change. I'll confirm and publish soon.

**Pull request merged**